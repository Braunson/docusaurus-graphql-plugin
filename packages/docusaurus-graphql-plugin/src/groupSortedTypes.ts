import {
  GraphQLNamedType,
  isEnumType,
  isInputObjectType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import { GroupedTypes, InterfaceItem, TypeItem } from "./types";

function groupTypes(types: GraphQLNamedType[]): GroupedTypes {
  return types.reduce<GroupedTypes>(
    (acc, type) => {
      if (type.name.startsWith("__")) {
        return acc;
      }

      if (isObjectType(type)) {
        if (type.name === "Query") {
          return {
            ...acc,
            queries: Object.values(type.getFields()).map((query) => ({
              type: query,
            })),
          };
        }

        if (type.name === "Mutation") {
          return {
            ...acc,
            mutations: Object.values(type.getFields()).map((mutation) => ({
              type: mutation,
            })),
          };
        }

        type.getInterfaces().forEach((inter) => {
          acc.interfaces.forEach((otherInter) => {
            if (otherInter.type.name === inter.name) {
              otherInter.implementedBy.push(type);
            }
          });
        });

        return {
          ...acc,
          objects: acc.objects.concat([{ type }]),
        };
      }

      if (isInterfaceType(type)) {
        const interfaceItem: InterfaceItem = {
          type,
          implementedBy: [],
        };

        acc.objects.forEach(({ type: object }) => {
          object.getInterfaces().forEach((otherInterface) => {
            if (otherInterface.name === interfaceItem.type.name) {
              interfaceItem.implementedBy.push(object);
            }
          });
        });

        return {
          ...acc,
          interfaces: acc.interfaces.concat([interfaceItem]),
        };
      }

      if (isEnumType(type)) {
        return {
          ...acc,
          enums: acc.enums.concat([{ type }]),
        };
      }

      if (isUnionType(type)) {
        return {
          ...acc,
          unions: acc.unions.concat([{ type }]),
        };
      }

      if (isInputObjectType(type)) {
        return {
          ...acc,
          inputObjects: acc.inputObjects.concat([{ type }]),
        };
      }

      if (isScalarType(type)) {
        return {
          ...acc,
          scalars: acc.scalars.concat([{ type }]),
        };
      }

      return acc;
    },
    {
      queries: [],
      mutations: [],
      objects: [],
      interfaces: [],
      enums: [],
      unions: [],
      inputObjects: [],
      scalars: [],
    }
  );
}

function sortByName<T extends TypeItem<{ name: string }>>(types: T[]): T[] {
  return types.sort((a, b) => a.type.name.localeCompare(b.type.name));
}

function sortGroupedTypes(groupedTypes: GroupedTypes): GroupedTypes {
  return {
    queries: sortByName(groupedTypes.queries),
    mutations: sortByName(groupedTypes.mutations),
    objects: sortByName(groupedTypes.objects),
    interfaces: sortByName(groupedTypes.interfaces),
    enums: sortByName(groupedTypes.enums),
    unions: sortByName(groupedTypes.unions),
    inputObjects: sortByName(groupedTypes.inputObjects),
    scalars: sortByName(groupedTypes.scalars),
  };
}

export function groupSortedTypes(types: GraphQLNamedType[]): GroupedTypes {
  return sortGroupedTypes(groupTypes(types));
}
