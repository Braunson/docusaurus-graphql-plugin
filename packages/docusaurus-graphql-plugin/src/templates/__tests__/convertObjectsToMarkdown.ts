import { buildSchema } from "graphql";
import { groupSortedTypes } from "../../groupSortedTypes";
import { convertObjectsToMarkdown } from "../convertObjectsToMarkdown";

describe("convertObjectsToMarkdown", () => {
  function getTypePath() {
    return "#";
  }

  it("should render objects", () => {
    const schema = buildSchema(`
      """
      A regular user.
      """
      type User {
        "The user's name."
        name: String!
      }
    `);
    const { objects } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertObjectsToMarkdown(objects, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });

  it("should list the object's interfaces", () => {
    const schema = buildSchema(`
      interface User {
        name: String!
      }

      """
      A user with special rights.
      """
      type Admin implements User {
        "The admin's name."
        name: String!
      }
    `);
    const { objects } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertObjectsToMarkdown(objects, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });
});
