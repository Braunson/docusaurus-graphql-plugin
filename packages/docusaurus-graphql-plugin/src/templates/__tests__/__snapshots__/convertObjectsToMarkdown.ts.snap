// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertObjectsToMarkdown should list the object's interfaces 1`] = `
"## Admin

A user with special rights.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implements</strong></p>

- [User](#)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The admin&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`convertObjectsToMarkdown should render objects 1`] = `
"## User

A regular user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;
